contracts/PreOrder.sol

```diff
@@ -0,0 +1,117 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+abstract contract PreOrder is Ownable {
+
+    struct Participant {
+        address sender;
+        uint amount;
+        uint time;
+    } 
+    
+    // the key `_nextPreOrder` is incremented with each additional pre-order
+    // make sure `_nextPreOrder` starts from 1 instead of 0
+    mapping(address => uint) private _preOrdered;
+    mapping(uint => Participant) private _preOrders;
+    uint private _nextPreOrder;
+    uint private _preOrderAmountTotal;
+
+    // determine whether there is an ongoing pre-order 
+    bool public inPreOrder;
+    // unit (wei)
+    uint public minAmount;
+    // participants allowed
+    uint256 public participantsAllowed;
+
+    error PreOrderNotStarted();
+    error PreOrderAlreadyStarted();
+    error PreOrderExists(address addr);
+    // minimum contribution amount allowed shall not be <= 0
+    error IncorrectMinAmount(uint amount);
+    error IncorrectParticipants(uint limit);
+    error MaxParticipantsReached(uint participants);
+
+    // start pre-order, only allowed by ower
+    function startPreOrder(uint amount, uint participants) public onlyOwner {
+        if (inPreOrder == true) revert PreOrderAlreadyStarted();
+        // min contribution shall always be larger than 0
+        if (amount <= 0) revert IncorrectMinAmount(amount);
```

Comment:
If we inherit from [`BatchNFT`](https://github.com/thematters/matters-avatar/blob/develop/contracts/BatchNFT.sol), maybe we can check if `amount` is smaller than `_totalSupply `?